### PR TITLE
[rollout] feat: support prompt embedding cache for diffusion rollouts

### PR DESCRIPTION
**Pull request by Sky-Trigger**
### What does this PR do?

This PR integrates vLLM-Omni's prompt embedding cache with verl-omni diffusion rollouts.

vLLM-Omni applies the cache wrapper around `encode_prompt()`. Previously, several
verl-omni adapters converted pre-tokenized prompt IDs and attention masks to tensors
before that boundary. Those tensor inputs bypassed the list-based cache-key path, so
repeated prompts were encoded again even when the cache was enabled.

This PR keeps two execution paths. With prompt caching disabled, the existing
tokenizer-to-tensor path is retained unchanged. With prompt caching enabled,
token IDs and masks remain list-like until `encode_prompt()`, where the existing
tensor conversion occurs after the wrapper has built its cache key. It covers:

- Qwen-Image FlowGRPO and MixGRPO (through the shared FlowGRPO adapter)
- Qwen-Image DiffusionNFT
- Qwen-Image online DPO
- Wan2.2 DanceGRPO

The adapters continue to accept explicit tensor inputs. We verified that both
paths preserve the output produced after `encode_prompt()` relative to the
pre-change implementation: prompt embeddings and masks retain the same values,
device, dtype, and batch/output contract. Precomputed prompt embeddings
continue to bypass token encoding.

The cache is disabled by default, so existing configurations do not enable it
implicitly.

For repeated rollouts, this PR also provides an opt-in routing-affinity mode.
When both the cache and affinity are enabled, all repeated generations derived
from one trainer sample use that sample's stable `uid` as the rollout request
ID. The load balancer can then route them to the same rollout replica, where
they share its worker-local cache. Without affinity, each generation uses a new
UUID and may be distributed across replicas; each replica then needs its own
first cache miss for the same prompt.

Related vLLM-Omni implementation: [vllm-project/vllm-omni#2962](https://github.com/vllm-project/vllm-omni/pull/2962).

### Checklist Before Starting

- [x] Search for similar PRs. Query: [prompt embedding cache PRs](https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+%22prompt+embed%22+cache)
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Test

The new CPU tests cover:

- Default tensor collation and opt-in list-preserving collation.
- Flat and packed/batched token-ID lists.
- List and tensor inputs at each adapter boundary.
- Preservation of tokenizer batch dimensions.
- Exact post-`encode_prompt()` parity with the pre-change output for both the
  cache-disabled tensor path and cache-enabled list path, including embeddings,
  masks, shape, dtype, and device.
- Cache hits for repeated list-valued Qwen-Image and Wan prompts.
- `prompt_ids=None` with precomputed prompt embeddings.
- DPO and DiffusionNFT request preparation without early tensorization.
- FlowGRPO forwarding list inputs through the cache boundary.
- Wan empty-negative-prompt handling for both list and tensor inputs.
- Configuration defaults and rejection of non-positive cache sizes.
- Cache-routing request IDs: random by default, stable per sample only when
  both prompt caching and routing affinity are enabled (covered by the separate
  agent-loop CPU tests).

Focused test command:

```bash
PYTHONPATH=/path/to/verl:/path/to/vllm-omni \
pytest -q \
  tests/pipelines/test_prompt_embed_cache_on_cpu.py \
  tests/workers/config/test_diffusion_config_on_cpu.py
```

Validation results:

- `git diff --check origin/main...HEAD`: passed.
- The focused suite completed with `46 passed` against the compatible local
  vLLM-Omni checkout used while developing the adapter changes.
- A later rerun against a different local vLLM-Omni revision stopped during test
  collection because that checkout no longer exports
  `vllm_omni.diffusion.worker.utils.DiffusionRequestState`. No test assertion was
  reached in that rerun; CI should use the repository's compatible pinned
  vLLM-Omni revision.

#### Performance evidence

The evaluation dataset used for the measurements below has short prompts, with
an average prompt length of only about 20--30 tokens. Text encoding therefore
accounts for a relatively small fraction of end-to-end generation and step time.
Based on the current comparison, this PR provides prompt-cache compatibility,
but the end-to-end gain is limited and it is not recommended as a priority
optimization. Workloads with longer prompts or more expensive text encoders may
benefit more, but that has not been measured here.

##### FlowGRPO

65b4ee316bda2ba3b22d5e42299b9876

##### DiffusionNFT

bb88eb87474cf9cfd3940bd45839db01

This DiffusionNFT plot is from the OCR dataset with each prompt repeated 10
times. In my assessment, the observed end-to-end improvement is small for this
workload. The repeated prompts do not make text encoding the dominant cost, and
replica-local cache routing limits how much of the repeated work can be reused.

These measurements do not by themselves establish the cache hit rate. Cache
entries are private to a rollout replica, so a repeated-prompt workload should
enable routing affinity when it needs repeated generations of one sample to
reuse the same replica-local entry.

### API and Usage Example

Enable the cache in the diffusion rollout configuration:

```yaml
actor_rollout_ref:
  rollout:
    enable_prompt_embed_cache: true
    prompt_embed_cache_size: 32
    enable_prompt_embed_cache_routing_affinity: true
```

Equivalent Hydra overrides:

```bash
actor_rollout_ref.rollout.enable_prompt_embed_cache=true \
actor_rollout_ref.rollout.prompt_embed_cache_size=32 \
actor_rollout_ref.rollout.enable_prompt_embed_cache_routing_affinity=true
```

`prompt_embed_cache_size` must be positive. The default configuration is:

```yaml
enable_prompt_embed_cache: false
prompt_embed_cache_size: 32
enable_prompt_embed_cache_routing_affinity: false
```

Routing affinity is meaningful only together with
`enable_prompt_embed_cache=true`. It does not change prompt values, sampling
parameters, or model weights; it only replaces the per-generation random
request ID with the repeated sample's `uid` for rollout-server routing.

#### Cache configuration

- Adds `enable_prompt_embed_cache` and `prompt_embed_cache_size` to
  `DiffusionRolloutConfig` and the generated trainer configurations.
- Validates that `prompt_embed_cache_size` is positive.
- Passes the cache enable and size options through `vLLMOmniHttpServer` to
  `AsyncOmni`.
- Adds `enable_prompt_embed_cache_routing_affinity`, disabled by default. The
  diffusion agent loop uses a sample's stable `uid` as the request ID only when
  it and prompt caching are both enabled; this is a verl-omni routing change,
  not an `AsyncOmni` cache-engine option. All other requests retain random
  UUIDs.

#### Prompt boundary

- Adds opt-in `preserve_lists=True` support to request-batch prompt-ID and mask
  collation. The default collation behavior remains tensor-based.
- Keeps pre-tokenized IDs and masks as lists through request preparation when they
  must participate in the prompt-cache key.
- Converts list inputs to device tensors only inside each adapter's or shared
  mixin's `encode_prompt()` implementation, after the cache wrapper has
  inspected the arguments.
- Retains the existing tensor path for callers that explicitly provide tensors.

#### Existing validated precedent: SD3.5 FlowGRPO

SD3.5 FlowGRPO already follows the same delayed-tensorization pattern in its
token-ID prompt-encoding boundary. Its `encode_prompt_from_token_ids()` path
accepts Python token-ID lists and converts them to device `torch.long` tensors
inside the CLIP and T5 encoding helpers, rather than before prompt encoding.

The existing `tests/pipelines/test_sd3_token_id_prompt_on_cpu.py` parity tests
compare this path with vLLM-Omni's text-based `encode_prompt()` and require exact
equality (`rtol=0`, `atol=0`). They also cover batch dimensions,
`num_images_per_prompt`, empty prompts, padding, and truncation. This provides an
already validated model-level precedent for moving list-to-tensor conversion into
the prompt-encoding boundary.

#### Adapter compatibility

- Preserves single-prompt and packed/batched prompt shapes.
- Computes batch size from either tensors or nested lists without hard-coding it.
- Preserves Qwen-Image True-CFG and Wan empty-negative-prompt behavior.
- Does not change the final prompt embedding or attention-mask contracts consumed
  by the denoising pipelines.

### CODE REVIEW

**Review by knlnguyen1802 — COMMENTED** [review]
Hi could you provide a convergent curse before review ?

**Comment by zhtmike** [comment]
need to update doc of how to use prompt cache correctly.

**Comment by Sky-Trigger** [comment]
@zhtmike @knlnguyen1802 
This is the FlowGRPO training curve before my PR is applied. Could you take a look and see if it looks right? There must be something wrong before this pr.
09e62166ca52716faefb0ba04d1f2a6b

**Comment by zhtmike** [comment]
gentle ping @knlnguyen1802

**Comment by knlnguyen1802** [comment]
> @zhtmike @knlnguyen1802 This is the FlowGRPO training curve before my PR is applied. Could you take a look and see if it looks right? There must be something wrong before this pr. 09e62166ca52716faefb0ba04d1f2a6b

I think we cant get any clue on this chart

File: verl_omni/pipelines/request_batch.py [hunk 1]
```diff
@@ -173,8 +221,15 @@ def collate_prompt_mask(
...
         field_name=field_name,
         pad_value=0,
+        preserve_lists=preserve_lists,
     )
     if mask is not None:
+        if isinstance(mask, list):
+            mask = [[value != 0 for value in row] for row in mask]
+            if target_seq_len is not None:
+                mask = [row[:target_seq_len] + [False] * max(target_seq_len - len(row), 0) for row in mask]
+            return mask
```
Comment (knlnguyen1802) [inline verl_omni/pipelines/request_batch.py:232]: Need to check carefully whether this padding is correct

File: verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py [hunk 2]
```diff
@@ -192,8 +207,7 @@ def _tokenize_text_prompt(self, text: str | list[str]):
             max_length=self.tokenizer_max_length + self.prompt_template_encode_start_idx,
             padding=True,
             truncation=True,
-            return_tensors="pt",
-        ).to(self.device)
```
Comment (knlnguyen1802) [inline verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py:196]: so this is redundant code until now ?

**Comment by Sky-Trigger** [comment]
the rewards chart does not 

> > @zhtmike @knlnguyen1802 This is the FlowGRPO training curve before my PR is applied. Could you take a look and see if it looks right? There must be something wrong before this pr. 09e62166ca52716faefb0ba04d1f2a6b
> 
> I think we cant get any clue on this chart

The rewards curve does not show the same normal upward trend as it did in PR #287. The curve shown here was obtained from a run on the current main branch without cherry-picking this PR.

**Comment by knlnguyen1802** [comment]
> The rewards curve does not show the same normal upward trend as it did in PR #287. The curve shown here was obtained from a run on the current main branch without cherry-picking this PR.

There is some problem with stepwise batching now, on analyze it

**Comment by Sky-Trigger** [comment]
> > The rewards curve does not show the same normal upward trend as it did in PR #287. The curve shown here was obtained from a run on the current main branch without cherry-picking this PR.
> 
> There is some problem with stepwise batching now, on analyze it

OK, I’ll show the speedup and the reward curve on DiffusionNFT using the OCR dataset with the prompt length manually increased from around 20 tokens to 1,000 tokens.

**Comment by Sky-Trigger** [comment]
@knlnguyen1802 Keeping both paths for now. The non-prompt-cache path is exactly the same as before. The prompt-cache path has been verified, and `encode_prompt` returns exactly the same as before. Updated the description above, feel free to take a look.

**Comment by Sky-Trigger** [comment]
ready for merge now

File: docs/contributing/integrating_prompt_embedding_cache.md
Comment (zhtmike) [inline docs/contributing/integrating_prompt_embedding_cache.md:1]: I think the doc is deviate from your implementation.

You change all pipelines for this, but the doc never mentions this

**Comment by zhtmike** [comment]
you should state clearly what need to be modified for support prompt caching in model contribution guide

**Comment by zhtmike** [comment]
and kindly ping @ZihaoW123 @AndyZhou952 @wtomin  for review for since they are the pipeline owners

File: docs/contributing/integrating_prompt_embedding_cache.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+- NumPy scalar values;
+- `list` and `tuple` containing only supported values;
+- `dict` whose values contain only supported values.
+
+`torch.Tensor`, PIL images, NumPy arrays, custom objects, and a list, tuple, or
+dictionary containing any of them are not cacheable. If any bound
+`encode_prompt()` argument is not cacheable, vLLM-Omni bypasses the cache for
+the entire call rather than building a partial key. Therefore, a list is not
+the only supported input type, but token IDs and masks should normally be kept
+as lists at the `encode_prompt()` boundary instead of tensors.
```
Comment (zhtmike) [inline docs/contributing/integrating_prompt_embedding_cache.md:59]: clear instructions. looks good

Comment (zhtmike) [reply docs/contributing/integrating_prompt_embedding_cache.md:59]: I think we should optimize the prompt cache hash in vllm-omni side, it is just too difficult to satisfy this criteria.
cc @SamitHuang @knlnguyen1802

Comment (Sky-Trigger) [reply docs/contributing/integrating_prompt_embedding_cache.md:59]: Yes, and this PR is mainly about making prompt caching compatible. But I don't actually recommend using it, since the performance gain is almost negligible compared to the potential risks.

**Review by zhtmike — APPROVED** [review]
looks good for me, waiting other's feedback

**Comment by ZihaoW123** [comment]
After making this change, have you tested whether WAN22 is working properly?

**Comment by Sky-Trigger** [comment]
> After making this change, have you tested whether WAN22 is working properly?

No, I haven’t run an end-to-end Wan2.2 test yet. However, I inspected the entire pipeline, and the adapter follows the same path as the other verified Qwen-Image RL algorithms. If this is a concern, I can remove the Wan2.2 adapter change from this patch and leave it for future contributors to address.

**Review by knlnguyen1802 — CHANGES_REQUESTED** [review]
For this PR, let's me verify it for QwenImage and SD3.5 on GPU first, recently we should take care of precision issue especially for QwenImage

**Comment by knlnguyen1802** [comment]
Update result running on GPU
@Sky-Trigger  Let's wait for verl pin update and move this PR forward
Also please verify on Wan2.2 or remove the adapter change
qwen_image_prompt_embedding_reward

**Comment by Sky-Trigger** [comment]
@knlnguyen1802 PTAL

**Review by knlnguyen1802 — APPROVED** [review]
LGTM now, make sure CI pass before merge

**Comment by Sky-Trigger** [comment]
@knlnguyen1802  all ci test passed
